server/jetstream_errors.go

```diff
@@ -0,0 +1,220 @@
+		JSStreamMsgDeleteFailedF:           {Code: 500, ErrCode: "JS057", Description: "%s"},
+		JSStreamNameExistErr:               {Code: 400, ErrCode: "JS058", Description: "stream name already in use"},
+		JSStreamNotFoundErr:                {Code: 404, ErrCode: "JS059", Description: "stream not found"},
+		JSStreamNotMatchErr:                {Code: 400, ErrCode: "JS060", Description: "expected stream does not match"},
+		JSStreamReplicasNotUpdatableErr:    {Code: 400, ErrCode: "JS061", Description: "Replicas configuration can not be updated"},
+		JSStreamRestoreErrF:                {Code: 500, ErrCode: "JS062", Description: "restore failed: %s"},
+		JSStreamSequenceNotMatchErr:        {Code: 503, ErrCode: "JS063", Description: "expected stream sequence does not match"},
+		JSStreamSnapshotErrF:               {Code: 500, ErrCode: "JS064", Description: "snapshot failed: %s"},
+		JSStreamSubjectOverlapErr:          {Code: 500, ErrCode: "JS065", Description: "subjects overlap with an existing stream"},
+		JSStreamTemplateCreateErrF:         {Code: 500, ErrCode: "JS066", Description: "%s"},
+		JSStreamTemplateDeleteErrF:         {Code: 500, ErrCode: "JS067", Description: "%s"},
+		JSStreamTemplateNotFoundErr:        {Code: 404, ErrCode: "JS068", Description: "template not found"},
+		JSStreamUpdateErrF:                 {Code: 500, ErrCode: "JS069", Description: "%s"},
+		JSStreamWrongLastMsgIDErrF:         {Code: 400, ErrCode: "JS070", Description: "wrong last msg ID: %s"},
+		JSStreamWrongLastSequenceErrF:      {Code: 400, ErrCode: "JS071", Description: "wrong last sequence: %d"},
+		JSTempStorageFailedErr:             {Code: 500, ErrCode: "JS072", Description: "JetStream unable to open temp storage for restore"},
+		JSTemplateNameNotMatchSubjectErr:   {Code: 400, ErrCode: "JS073", Description: "template name in subject does not match request"},
+	}
+)
+
+type natsCodedError interface {
+	natsErrCode() string
+}
+
+func IsNatsCodedErr(err error, code string) bool {
+	if err == nil {
+		return false
+	}
+
+	ce, ok := err.(natsCodedError)
+	if !ok {
+		return false
+	}
+
+	return ce.natsErrCode() == code
+}
+
+// ApiError is included in all responses if there was an error.
+// TODO(dlc) - Move to more generic location.
+type ApiError struct {
+	Code        int    `json:"code"`
+	ErrCode     string `json:"err_code,omitempty"`
+	Description string `json:"description,omitempty"`
+	URL         string `json:"-"`
+	Help        string `json:"-"`
+}
+
+func (e *ApiError) natsErrCode() string { return e.ErrCode }
+
+func (e *ApiError) Error() string {
+	return e.Description + " (" + e.ErrCode + ")"
+}
+
+func (e *ApiError) NewF(a ...interface{}) *ApiError {
+	return &ApiError{
+		Code:        e.Code,
+		ErrCode:     e.ErrCode,
+		Description: fmt.Sprintf(e.Description, a...),
+	}
+}
+
+func (e *ApiError) ErrOrNewF(err error, a ...interface{}) *ApiError {
```

Comment:
will add some comments

Reply:
I am the worst at naming but might consider something a bit more descriptive.

Reply:
I'd need some suggestions, as these are already attempts at improving the situation :)

Reply:
Describe the function for us.

Reply:
Uncommited locally I have this comment

```
// ErrOrNewF returns err if it's an ApiError else creates a new error using Sprintf
```

Reply:
```
err = acc.deleteStreamTemplate(name)
if err != nil {
		resp.Error = ApiErrors[JSStreamTemplateDeleteErrF].ErrOrNew(err)
		s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
		return
}
```

Here if deleteStreamTemplate() returns an apiError - that one will be used, otherwise a new `JSStreamTemplateDeleteErrF` is made via Sprintf

This way we can return a ApiError for things like template not found but a io error if deleting it from disk fails, and the fact that the problem is template not found would be surfaced correctly.